lib/lattice_field.cpp

```diff
@@ -236,32 +240,46 @@ namespace quda {
           qudaDeviceSynchronize();
           comm_barrier();
           for (int b=0; b<2; b++) {
-	    device_pinned_free(ghost_recv_buffer_d[b]);
-	    device_pinned_free(ghost_send_buffer_d[b]);
-	    host_free(ghost_pinned_send_buffer_h[b]);
-	    host_free(ghost_pinned_recv_buffer_h[b]);
-	  }
+#ifndef NVSHMEM_COMMS
+            device_pinned_free(ghost_recv_buffer_d[b]);
+            device_pinned_free(ghost_send_buffer_d[b]);
+#else
+            shmem_pinned_free(ghost_recv_buffer_d[b]);
+            shmem_pinned_free(ghost_send_buffer_d[b]);
+#endif
+            host_free(ghost_pinned_send_buffer_h[b]);
+            host_free(ghost_pinned_recv_buffer_h[b]);
+          }
         }
       }
 
       if (ghost_bytes > 0) {
         for (int b = 0; b < 2; ++b) {
           // gpu receive buffer (use pinned allocator to avoid this being redirected, e.g., by QDPJIT)
-	  ghost_recv_buffer_d[b] = device_pinned_malloc(ghost_bytes);
-
-	  // gpu send buffer (use pinned allocator to avoid this being redirected, e.g., by QDPJIT)
-	  ghost_send_buffer_d[b] = device_pinned_malloc(ghost_bytes);
-
-	  // pinned buffer used for sending
-	  ghost_pinned_send_buffer_h[b] = mapped_malloc(ghost_bytes);
-
-	  // set the matching device-mapped pointer
-	  ghost_pinned_send_buffer_hd[b] = get_mapped_device_pointer(ghost_pinned_send_buffer_h[b]);
-
-	  // pinned buffer used for receiving
-	  ghost_pinned_recv_buffer_h[b] = mapped_malloc(ghost_bytes);
-
-	  // set the matching device-mapped pointer
+#ifndef NVSHMEM_COMMS
+          ghost_recv_buffer_d[b] = device_pinned_malloc(ghost_bytes);
+#else
+          ghost_recv_buffer_d[b] = shmem_pinned_malloc(ghost_bytes);
+#endif
+
+          // gpu send buffer (use pinned allocator to avoid this being redirected, e.g., by QDPJIT)
+          // /MWTODO: This is a NVSHMEM_IPC_HACK
+#ifndef NVSHMEM_COMMS
+          ghost_send_buffer_d[b] = device_pinned_malloc(ghost_bytes);
+#else
+          ghost_send_buffer_d[b] = shmem_pinned_malloc(ghost_bytes);
+#endif
+
+          // pinned buffer used for sending
+          ghost_pinned_send_buffer_h[b] = mapped_malloc(ghost_bytes);
+
+          // set the matching device-mapped pointer
+          cudaHostGetDevicePointer(&ghost_pinned_send_buffer_hd[b], ghost_pinned_send_buffer_h[b], 0);
```

Comment:
this needs to be reverted back to 
`ghost_pinned_send_buffer_hd[b] = get_mapped_device_pointer(ghost_pinned_send_buffer_h[b]);`